Rac dropzone/filetrigger docs (#4847)
adobe/react-spectrum: packages/react-aria-components/docs/DropZone.mdx

    --text-color: ButtonText;
    --focus-ring-color: Highlight;
  }
}
```

</details>

## Features

There is no native element to implement a dropzone in HTML that is widely supported. `DropZone` helps achieve accessible dropzone components that can be styled as needed. 

* **Styleable** – Hover, keyboard focus, and the drop target state is provided for easy styling. These styles only apply when interacting with an appropriate input device, unlike CSS pseudo classes. 
* **Accessible** – Uses a visually hidden `<button>` under the hood to allow interactions like dropping objects with a keyboard. 
* **Flexible** – Can be used standalone or with `FileTrigger` to allow file uploads.

## Anatomy

A dropzone consists of a target element for the dropped objects. Users may drop objects via mouse or keyboard. In addition, a `FileTrigger` may be passed as a child to allow access to computer's file system. 

If a visual label is not provided, then an `aria-label` or
`aria-labelledby` prop must be passed to identify the button to assistive technology.

### Composed Components
A dropzone can pass `FileTrigger` as a child, which may also be used standalone or reused in other components.

> {/* TODO: Add an Example Card of FileTrigger */}

Author: There aren't any svg's to go along with DropZone or FileTrigger's example cards so I figured we just leave it out for now.

  width: 25%;
  display: inline-block;

  &[data-drop-target] {
    box-shadow: 0 0 0 1px var(--focus-ring-color);
    background: var(--background-color-pressed);
    border-color: var(--border-color-pressed);
  }

  &[data-focus-visible] {
    border-color: var(--focus-ring-color);
    box-shadow: 0 0 0 1px var(--focus-ring-color);
  }
}

@media (forced-colors: active) {
  .react-aria-DropZone {
    forced-color-adjust: none;
    --border-color: ButtonBorder;
    --border-color-drop-target: ButtonBorder;
    --background-color: ButtonFace;
    --background-color-pressed: ButtonFace;
    --text-color: ButtonText;
    --focus-ring-color: Highlight;
  }
}
```

</details>

## Features

There is no native element to implement a dropzone in HTML that is widely supported. `DropZone` helps achieve accessible dropzone components that can be styled as needed. 

* **Styleable** – Hover, keyboard focus, and the drop target state is provided for easy styling. These styles only apply when interacting with an appropriate input device, unlike CSS pseudo classes. 
* **Accessible** – Uses a visually hidden `<button>` under the hood to allow interactions like dropping objects with a keyboard. 

> * **Flexible** – Can be used standalone or with `FileTrigger` to allow file uploads.

Reviewer: I would call out the support for various data types too. FileTrigger is technically a separate component, and you can render whatever you want inside a drop zone.

```suggestion
* **Flexible** – Files, directories, and custom data types can be dropped, and the contents of the drop zone can be fully customized.
```

## Anatomy

> A dropzone consists of a target element for the dropped objects. Users may drop objects via mouse, keyboard, or touch. In addition, a `FileTrigger` may be passed as a child to allow users to upload files. 

Reviewer: ```suggestion
A drop zone consists of a target element for the dropped objects. Users may drop objects via mouse, keyboard, or touch. DropZone accepts any content as its children, which may change when the user drops content. A [FileTrigger](FileTrigger.html) is commonly paired with a DropZone to allow a user to choose files from their device.
```

If a visual label is not provided, then an `aria-label` or

> `aria-labelledby` prop must be passed to identify the visually hidden button to assistive technology.

Reviewer: We might want to add something about the Text slot here as well since that is also used as the label?

### Composed Components

> A dropzone can pass `FileTrigger` as a child, which may also be used standalone or reused in other components.

Reviewer: ```suggestion
A drop zone can include a `FileTrigger` as a child, which may also be used standalone or reused in other components.
```

```css
.react-aria-DropZone {
  /* ... */
}
```

A custom `className` can also be specified on any component. This overrides the default `className` provided by React Aria with your own.

```jsx
<DropZone className="my-dropzone">
  {/* ... */}
</DropZone>
```

In addition, some components support multiple UI states (e.g. focused, placeholder, readonly, etc.). React Aria components expose states using DOM attributes, which you can target in CSS selectors. These are ARIA attributes wherever possible, or data attributes when a relevant ARIA attribute does not exist. For example:

```css
.react-aria-DropZone[data-drop-target] {
  /* ... */
}
```

The `className` and `style` props also accept functions which receive states for styling. This lets you dynamically determine the classes or styles to apply, which is useful when using utility CSS libraries like [Tailwind](https://tailwindcss.com/).

```jsx
<DropZone className={({isDropTarget}) => isDropTarget ? 'bg-gray-700' : 'bg-gray-600'} />
```

> Render props may also be used as children to alter what elements are rendered based on the current state. For example, you could render an extra element when the button is in a pressed state.

Reviewer: ```suggestion
Render props may also be used as children to alter what elements are rendered based on the current state. For example, you could render an extra element when the drop target is in an active state.
```

        }}>
        <Text slot="heading">
          {dropped || 'Drop here'}
        </Text>
      </ DropZone>
    </>
  );
}

<Example />

```

#### Draggable

The `Draggable` component used above is defined below. See [useDrag](useDrag.html) for more details and documentation.

<details>
  <summary style={{fontWeight: 'bold'}}><ChevronRight size="S" /> Show code</summary>

```tsx example render=false export=true
import {useDrag} from '@react-aria/dnd';

function Draggable() {
  let {dragProps, isDragging} = useDrag({
    getItems() {
      return [{
        'text/plain': 'hello world',
        'my-app-custom-type': JSON.stringify({message: 'hello world'})
      }];
    }
  });

  return (
    <div {...dragProps} role="button" tabIndex={0} className={`draggable ${isDragging ? 'dragging' : ''}`}>
      Drag me
    </div>
  );
}
```
</details>

<details>
  <summary style={{fontWeight: 'bold'}}><ChevronRight size="S" /> Show CSS</summary>

```css
.draggable {
  display: inline-block;
  vertical-align: top;
  border: 1px solid gray;
  padding: 10px;
  margin-right: 20px;
  border-radius: 4px;
}

.draggable.dragging {
  opacity: 0.5;
}
```

</details>

### Heading

> The `heading` slot enables the user to reference the text content to define the dropzone's accessible name. 

Reviewer: Seeing this here I think we should probably rename the `heading` slot to `label` since it isn't on a `<Heading>` element. We seem to use "label" in a bunch of other components already.

```tsx example
import {Text} from 'react-aria-components';

<DropZone>
  <Text slot="heading">
    Drop item here
  </Text>
</DropZone>
```

> ### File Upload

Reviewer: ```suggestion
### FileTrigger
```

> To upload files, pass `FileTrigger` as a child of `DropZone`.

Reviewer 2: I would probably remove mentions of upload from the docs, since this is just a client-side thing.

Maybe something like this:

```suggestion
To allow selection of files from the user's device, pass `FileTrigger` as a child of `DropZone`.
```

  --focus-ring-color: slateblue;

  color: var(--text-color);
  background: var(--background-color);
  border: 1px solid var(--border-color);
  border-radius: 4px;
  appearance: none;
  vertical-align: middle;
  font-size: 1.2rem;
  text-align: center;
  margin: 0;
  outline: none;
  padding: 4px 12px;

  &[data-pressed] {
    box-shadow: inset 0 1px 2px rgb(0 0 0 / 0.1);
    background: var(--background-color-pressed);
    border-color: var(--border-color-pressed);
  }

  &[data-focus-visible] {
    border-color: var(--focus-ring-color);
    box-shadow: 0 0 0 1px var(--focus-ring-color);
  }

  &:disabled {
    border-color: var(--border-color-disabled);
    color: var(--text-color-disabled);
  }
}

@media (forced-colors: active) {
  .react-aria-Button {
    forced-color-adjust: none;
    --border-color: ButtonBorder;
    --border-color-pressed: ButtonBorder;
    --border-color-disabled: GrayText;
    --background-color: ButtonFace;
    --background-color-pressed: ButtonFace;
    --text-color: ButtonText;
    --text-color-disabled: GrayText;
    --focus-ring-color: Highlight;
  }
}
```

</details>

### Visual feedback

> A dropzone can provide visual feedback to the user when a drag hovers over the drop target by passing the `getDropOpertation` function. If a drop target only supports data of specific types (e.g. images, videos, text, etc.), then it should implement getDropOperation and return 'cancel' for types that aren't supported. This will prevent visual feedback indicating that the drop target accepts the dragged data when this is not true. [Read more about getDropOperation.](https://react-spectrum.adobe.com/react-aria/useDrop.html#getdropoperation)

Reviewer: ```suggestion
A dropzone displays visual feedback to the user when a drag hovers over the drop target. If a drop target only supports data of specific types (e.g. images, videos, text, etc.), then it should implement the `getDropOperation` prop and return `'cancel'` for types that aren't supported. This will prevent visual feedback indicating that the drop target accepts the dragged data when this is not true. [Read more about getDropOperation.](https://react-spectrum.adobe.com/react-aria/useDrop.html#getdropoperation)
```

I think eventually we can also add the `acceptedDragTypes` prop to make this a bit easier, but that is out of scope for this PR.

Author: oh yeah i believe we have a [follow-up ticket](https://github.com/orgs/adobe/projects/19/views/4?filterQuery=dropzone&pane=issue&itemId=31210918) for that already

```tsx example
<DropZone

>   getDropOperation={(types) =>  types.has('image/png') ? 'copy' : 'cancel'} >

Reviewer: ```suggestion
  getDropOperation={(types) => types.has('image/png') ? 'copy' : 'cancel'}>
```

  Drop files here
</DropZone>
```

> 

Reviewer: I think we should probably copy more of the content from the `useDrop` page into here at some point. For example, it covers how to work with various data types such as files, directories, and custom types. It also has a bit more description about drop operations, e.g. what they are, the various cursors etc. But let's do that as a separate PR.

Author: cool i'll make a follow-up ticket for that

export default Layout;

import docs from 'docs:react-aria-components';
import typesDocs from 'docs:@react-types/shared/src/events.d.ts';
import {PropTable, HeaderInfo, TypeLink, PageDescription, StateTable} from '@react-spectrum/docs';
import styles from '@react-spectrum/docs/src/docs.css';
import packageData from 'react-aria-components/package.json';
import Anatomy from '@react-aria/datepicker/docs/daterangepicker-anatomy.svg';
import ChevronRight from '@spectrum-icons/workflow/ChevronRight';
import {Divider} from '@react-spectrum/divider';
import {ExampleCard} from '@react-spectrum/docs/src/ExampleCard';
import {Keyboard} from '@react-spectrum/text';

---
category: Drag and drop
keywords: [dropzone, drop, dnd, drag and drop, aria, accessibility]
type: component
---

# DropZone

<PageDescription>{docs.exports.DropZone.description}</PageDescription>

<HeaderInfo
  packageData={packageData}
  componentNames={['DropZone']} />

## Example

```tsx example
import {DropZone, Text} from 'react-aria-components';

function Example() {
  let [dropped, setDropped] = React.useState(false);

  return (
    <DropZone
      onDrop={() => {
        setDropped(true);
      }}>

>       {dropped ? "You dropped something" : "Drop object here"}

Reviewer: Did you mean to wrap this in `<Text slot="label">`? I see it's imported above.

Author: oops yeah
